src/VisualStudio/Xaml/Impl/Diagnostics/Analyzers/XamlDocumentDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,86 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Xaml;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Package = Microsoft.VisualStudio.Shell.Package;
+
+namespace Microsoft.CodeAnalysis.Xaml.Diagnostics.Analyzers
+{
+    [DiagnosticAnalyzer(StringConstants.XamlLanguageName)]
+    internal class XamlDocumentDiagnosticAnalyzer : DocumentDiagnosticAnalyzer
+    {
+        private IXamlDocumentAnalyzerService analyzerService;
+
+        private IXamlDocumentAnalyzerService AnalyzerService
+        {
+            get
+            {
+                if (this.analyzerService == null)
+                {
+                    IComponentModel componentModel = (IComponentModel)Package.GetGlobalService(typeof(SComponentModel));
+                    if (componentModel == null)
+                    {
+                        Debug.Fail("Missing IComponentModel!");
+                        return null;
+                    }
+
+                    this.analyzerService = componentModel.GetExtensions<IXamlDocumentAnalyzerService>().FirstOrDefault();
+                    Debug.Assert(this.analyzerService != null, "Missing IXamlDocumentAnalyzerService!");
+                }
+
+                return this.analyzerService;
+            }
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return this.AnalyzerService?.SupportedDiagnostics ?? ImmutableArray.Create<DiagnosticDescriptor>();
+            }
+        }
+
+        public override async Task<ImmutableArray<Diagnostic>> AnalyzeSyntaxAsync(Document document, CancellationToken cancellationToken)
+        {
+            if (document.Project.Solution.Workspace.Kind != WorkspaceKind.Host)
+            {
+                // Don't bother supporting diagnostics in non-host workspaces.  For example, we 
+                // don't want to bother computing diagnostics for 'Preview' workspaces.  That's
+                // a whole lot of work for a tiny preview.
+                return ImmutableArray<Diagnostic>.Empty;
+            }
+
+            if (this.AnalyzerService == null)
+            {
+                return ImmutableArray<Diagnostic>.Empty;
+            }
+
+            return await this.AnalyzerService.AnalyzeSyntaxAsync(document, cancellationToken).ConfigureAwait(false);
+        }
+
+        public override async Task<ImmutableArray<Diagnostic>> AnalyzeSemanticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            if (document.Project.Solution.Workspace.Kind != WorkspaceKind.Host)
+            {
+                // Don't bother supporting diagnostics in non-host workspaces.  For example, we 
+                // don't want to bother computing diagnostics for 'Preview' workspaces.  That's
+                // a whole lot of work for a tiny preview.
+                return ImmutableArray<Diagnostic>.Empty;
```

Comment:
Note that we *do* compute diagnostics for C# and VB for the preview, which is kind of nice, since you can see if the action you are performing introduces something else broken.